File: circuitpython-code/state_machine.py
"""
The MIT License (MIT)

Copyright (c) 2020 Dave Astels

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

--------------------------------------------------------------------------------
The satet machine
"""

import adafruit_logging as logging
logger = logging.getLogger('mouse')

class StateMachine(object):

    def __init__(self, system, name):
        self._system = system
        self._name = name
        self._state = None
        self._states = {}
        self._initial_state = None


    def dispose(self):
        self._states.clear()


    def add_state(self, state):
        self._states[state.name] = state


    @property
    def initial_state(self):
        return self._initial_state


    def reset(self):
        self.go_to_state(self._initial_state)


    def go_to_state(self, state_name, data=None):
        if state_name:
            if self._state:
                logger.info('Exiting %s', self._state)
                self._state.exit(self)
            self._state = self._states[state_name]
            logger.info('Entering %s', self._state)
            self._state.enter(data)
        else:
            logger.critical('None state name')
            raise Exception('None state name')


    def update(self, now):
        if self._state:
 #           logger.debug('Updating %s', self._state)
            self._state.update(now)


    def event_occurred(self, event):
        if self._state:
            logger.debug('Event in %s: %s', self._state, str(event))
            self._state.event_occurred(event, self)


    @property
    def current_state(self):
        return self._state.name if self._state else 'NOT_SET'


    @property
    def name(self):
        return self._name

    @property
    def system(self):
        return self._system


    def __str__(self):
        return '<State Machine: {0}>'.format(self._name)

File: c++-code/src/state_machine.cpp
// State Machine base class
//
// Copyright (c) 2020 Dave Astels

#include <string.h>
#include "logging.h"
#include "state_machine.h"

extern Logger *logger;

StateMachine::StateMachine(System *system, const char *name, bool should_log_transitions)
  : _system(system)
  , _name(name)
  , _current_state(NULL)
  , _number_of_states(0)
  , _log_transitions(should_log_transitions)
{
}


bool StateMachine::add_state(State *state)
{
  if (_number_of_states == MAX_STATES) {
    logger->critical("Adding state %s to %s failed", state->name(), _name);
    return false;
  }
  _states[_number_of_states++] = state;
  return true;
}


bool StateMachine::reset()
{
  if (go_to_state(_initial_state_name)) {
    return true;
  }
  logger->critical("Resetting %s to %s failed!", _name, _initial_state_name);
  return false;
}


State *StateMachine::find_state(const char *state_name)
{
  for (int index = 0; index < _number_of_states; index++) {
    if (_states[index]->is_named(state_name)) {
      return _states[index];
    }
  }
  return nullptr;
}


bool StateMachine::go_to_state(const char *state_name, void *data)
{
  if (!state_name || !*state_name) { // no state name
    logger->critical("No state name to go to in %s", _name);
    return false;
  }
  State *new_state = find_state(state_name);
  if (new_state == nullptr) {   // bad state name
    logger->critical("Tried to go to bad state %s in %s", state_name, _name);
    return false;
  }
  if (_current_state) {
    if (_log_transitions) logger->debug("Exiting %s:%s", name(), _current_state->name());
    _current_state->exit();
  }
  _current_state = new_state;
  if (_log_transitions) logger->debug("Entering %s:%s", name(), _current_state->name());
  _current_state->enter(data);
  return true;
}


void StateMachine::update(uint32_t now)
{
  if (_current_state) {
    _current_state->update(now);
  }
}


void StateMachine::event_occurred(Event *event)
{
  if (_current_state) {
    _current_state->event_occurred(event);
  }
}


const char *StateMachine::current_state_name()
{
  if (_current_state) {
    return _current_state->name();
  } else {
    return "NOT_SET";
  }
}
